### PR TITLE
Change bounds on `TryFrom` blanket impl to use `Into` instead of `From`

### PR DESCRIPTION
**Pull request by RustyYato**
This is from this [comment](https://github.com/rust-lang/rust/issues/33417#issuecomment-447111156) I made.

This will expand the impls available for `TryFrom` and `TryInto`, without losing anything in the process.

### CODE REVIEW

**Comment by KrishnaSannasi** [comment]
@TimNN I can't figure out where it thinks there is an impl conflict, could you please help me find it?

**Comment by shepmaster** [comment]
> thinks there is an impl conflict

There's a test that's asserting on an error message and your changes have changed the error message:

> [00:47:37] ---- [ui] ui/e0119/conflict-with-std.rs stdout ----
> [00:47:37] diff of stderr:

**Comment by SimonSapin** [comment]
I *think* this is fine, especially since `TryFrom` is unstable. @rust-lang/libs, any other opinion?

**Comment by sfackler** [comment]
This seems weird to me, honestly.

**Comment by KrishnaSannasi** [comment]
@sfackler Why?

**Comment by sfackler** [comment]
Because these are direct promotions of the related trait, not the opposite version of that trait.

**Comment by KrishnaSannasi** [comment]
@sfackler I understand that, but if we want the impl `TryFrom` to imply `TryInto`, which I think we do, and we want to keep the impl `From` to imply `Into`, then the only way to get `Into` to imply `TryInto` is by going through `TryFrom`. Otherwise we could get conflicting impls of `TryInto`. This is simply a way to sidestep that. It does look a bit weird, but it is necessary to have the obvious things happen.

We can think of it like this
`From` implies `Into` (and these two traits are semantically equivalent)
`Into` implies `TryFrom` (this impl stitches together fallible and infallible conversions)
`TryFrom` implies `TryInto` (similar to how `From` implies `Into`, `TryFrom` and `TryInto` are semantically equivalent, so this impl must exist in one form or another)

Another way to do it would be to do
`From` implies `Into`
`Into` implies `TryInto`
`TryInto` implies `TryFrom`

This would also work, and may be easier to think about.

---
If we don't want to change the `TryFrom` implies `TryInto` impl, then the only way to get all infallible conversions to have fallible equivalents (where the `Err` is `!`), is to have `Into` imply `TryFrom`.
Or we could chave `Into` imply `TryInto` and `TryInto` imply `TryFrom`, and that would pipeline all conversions.

---
Overall, having a tree structure for this would be bad, as it would punish some impls for reasons beyond their control. So I want to have a linear structure.

Tree:
From -> TryFrom -> TryInto
L> Into

Linear:
From -> Into -> TryFrom -> TryInto
or
From -> Into -> TryInto -> TryFrom

**Comment by KrishnaSannasi** [comment]
Are there any more objections?

**Comment by KrishnaSannasi** [comment]
@shepmaster

I think all objections have been dealt with, can this be merged now?

**Comment by shepmaster** [comment]
@bors r+ rollup

I think, since it's unstable, it's worth giving this a shot.

@KrishnaSannasi is it possible to write a test that shows the benefit of this ordering that would not be possible with the current setup?

**Comment by KrishnaSannasi** [comment]
I'm not sure how to write tests, as this is my first go at hacking into the compiler, could you please give an example of how to do it?

**Comment by KrishnaSannasi** [comment]
@shepmaster I added the test, will that be sufficient or should it be different?

**Comment by Centril** [comment]
Assigning over to make sure that
r? @sfackler 
agrees with merging this.

File: src/test/run-pass/try_from.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This test relies on `TryFrom` being auto impl for all `T: Into`
+// and `TryInto` being auto impl for all `U: TryFrom`
```
Comment (shepmaster) [inline src/test/run-pass/try_from.rs]: I'd rephrase this as "blanket implemented"  instead of "auto impl" as `auto trait` is a separate concept.

File: src/test/run-pass/try_from.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+use std::convert::TryInto;
+
+struct Foo<T> {
+    t: T
+}
+
+/*
+// This fails to compile due to coherence restrictions
+// as of rust version 1.32.x
```
Comment (shepmaster) [inline src/test/run-pass/try_from.rs]: Minor nit, Rust is a proper noun and should be capitalized.

File: src/test/run-pass/try_from.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+impl<T> From<Foo<T>> for Box<T> {
+    fn from(foo: Foo<T>) -> Box<T> {
+        Box::new(foo.t)
+    }
+}
```
Comment (shepmaster) [inline src/test/run-pass/try_from.rs]: Is this a test that *should* compile with the code in this PR? If so, why is it commented out? Otherwise, what value does it add being here?

File: src/test/run-pass/try_from.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
```
Comment (shepmaster) [inline src/test/run-pass/try_from.rs]: Double check some surrounding test files; I *think* these license headers are in the process of being removed...

File: src/test/run-pass/try_from.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+// This test was added to show the motivation for doing this
+// over `TryFrom` being auto impl for all `T: From`
+
+#![feature(try_from, never_type)]
+
+use std::convert::TryInto;
+
+struct Foo<T> {
+    t: T
```
Comment (shepmaster) [inline src/test/run-pass/try_from.rs]: Pedantic: add a trailing comma here.

File: src/test/run-pass/try_from.rs (hunk 3 above)
Comment (KrishnaSannasi) [reply src/test/run-pass/try_from.rs]: I added it to show why we needed to use the `Into` version instead of the `From` version of the impl, but if it is unneeded, I can take it out

**Comment by shepmaster** [comment]
@sfackler any further objections?
